### PR TITLE
Implement asynchronous support in ODataJsonLightParameterDeserializer

### PR DESCRIPTION
**Pull request by gathogojr**
### Issues

This pull request partially fulfills #2019.

### Description

Implement asynchronous support in **`ODataJsonLightParameterDeserializer`**.

### Checklist (Uncheck if it is not completed)

- [x] *Test cases added*
- [ ] *Build and test with one-click build and test script passed*

### Additional work necessary

*If documentation update is needed, please add "Docs Needed" label to the issue and provide details about the required document change in the issue.*

### CODE REVIEW

File: test/FunctionalTests/Microsoft.OData.Core.Tests/JsonLight/ODataJsonLightParameterDeserializerTests.cs [hunk 1]
```diff
@@ -0,0 +1,392 @@
...
+
+            this.InitializeModel();
+            this.propertyAndAnnotationCollector = new PropertyAndAnnotationCollector(true);
+        }
+
+        [Fact]
+        public async Task ReadPrimitiveParameterAsync()
+        {
+            var payload = "{\"rating\":4}";
+            var setRatingAction = new EdmAction("NS", "SetRating", null);
```
Comment (marabooy) [inline test/FunctionalTests/Microsoft.OData.Core.Tests/JsonLight/ODataJsonLightParameterDeserializerTests.cs:40]: nit: Move this closer to usage.

File: src/Microsoft.OData.Core/JsonLight/ODataJsonLightParameterDeserializer.cs [hunk 2]
```diff
@@ -216,5 +222,192 @@ internal bool ReadNextParameter(PropertyAndAnnotationCollector propertyAndAnnota
...
+            this.AssertJsonCondition(JsonNodeType.Property, JsonNodeType.EndObject);
+
+            bool parameterRead = false;
+            if (this.JsonReader.NodeType == JsonNodeType.Property)
+            {
+                bool foundCustomInstanceAnnotation = false;
+                await this.ProcessPropertyAsync(
+                    propertyAndAnnotationCollector,
+                    propertyAnnotationValueReaderAsync,
+                    async (propertyParsingResult, parameterName) =>
```
Comment (marabooy) [inline src/Microsoft.OData.Core/JsonLight/ODataJsonLightParameterDeserializer.cs:251]: A good candidate for a future update on this lambda so it captures less items. :smile:

File: test/FunctionalTests/Microsoft.OData.Core.Tests/JsonLight/ODataJsonLightParameterDeserializerTests.cs (hunk 1 above)
Comment (gathogojr) [reply test/FunctionalTests/Microsoft.OData.Core.Tests/JsonLight/ODataJsonLightParameterDeserializerTests.cs:39]: Fixed

File: src/Microsoft.OData.Core/JsonLight/ODataJsonLightParameterDeserializer.cs (hunk 2 above)
Comment (gathogojr) [reply src/Microsoft.OData.Core/JsonLight/ODataJsonLightParameterDeserializer.cs:251]: I plan to tackle in a separate to address the same concern in related PRs

**Review by habbes — APPROVED** [review]

File: src/Microsoft.OData.Core/JsonLight/ODataJsonLightParameterDeserializer.cs [hunk 3]
```diff
@@ -25,6 +26,11 @@ internal sealed class ODataJsonLightParameterDeserializer : ODataJsonLightProper
         private static readonly Func<string, object> propertyAnnotationValueReader =
             annotationName => { throw new ODataException(ODataErrorStrings.ODataJsonLightParameterDeserializer_PropertyAnnotationForParameters); };
 
+        /// <summary>OData property annotation asynchronous reader for parameter payloads.</summary>
```
Comment (ElizabethOkerio) [inline src/Microsoft.OData.Core/JsonLight/ODataJsonLightParameterDeserializer.cs:29]: I don't think this statement is necessary here. Its contradicting the second statement. You can get rid of this and just leave the second one.

**Review by ElizabethOkerio — APPROVED** [review]

File: src/Microsoft.OData.Core/JsonLight/ODataJsonLightParameterDeserializer.cs [hunk 4]
```diff
@@ -216,5 +222,192 @@ internal bool ReadNextParameter(PropertyAndAnnotationCollector propertyAndAnnota
...
+        /// Pre-Condition:  Property or EndObject   the property node of the parameter to read or the end object node if there are not parameters
+        /// Post-Condition: Property or EndObject   the node after the property value of a primitive, complex or null collection parameter
+        ///                 Any                     the start of the value representing a non-null collection parameter (the collection reader will fail if this is not a StartArray node)
+        /// </remarks>
+        internal async Task<bool> ReadNextParameterAsync(PropertyAndAnnotationCollector propertyAndAnnotationCollector)
+        {
+            Debug.Assert(propertyAndAnnotationCollector != null, "propertyAndAnnotationCollector != null");
+            this.AssertJsonCondition(JsonNodeType.Property, JsonNodeType.EndObject);
+
+            bool parameterRead = false;
```
Comment (KenitoInc) [inline src/Microsoft.OData.Core/JsonLight/ODataJsonLightParameterDeserializer.cs:244]:
```suggestion
            bool parameterRead = false;

```

File: src/Microsoft.OData.Core/JsonLight/ODataJsonLightParameterDeserializer.cs [hunk 5]
```diff
@@ -216,5 +222,192 @@ internal bool ReadNextParameter(PropertyAndAnnotationCollector propertyAndAnnota
...
+                            case PropertyParsingResult.MetadataReferenceProperty:
+                                throw new ODataException(ODataErrorStrings.ODataJsonLightPropertyAndValueDeserializer_UnexpectedMetadataReferenceProperty(parameterName));
+
+                            case PropertyParsingResult.PropertyWithValue:
+                                IEdmTypeReference parameterTypeReference = this.parameterReader.GetParameterTypeReference(parameterName);
+                                Debug.Assert(parameterTypeReference != null, "parameterTypeReference != null");
+
+                                ODataParameterReaderState state;
+                                object parameterValue;
+                                EdmTypeKind parameterTypeKind = parameterTypeReference.TypeKind();
```
Comment (KenitoInc) [inline src/Microsoft.OData.Core/JsonLight/ODataJsonLightParameterDeserializer.cs:287]:
```suggestion
                                EdmTypeKind parameterTypeKind = parameterTypeReference.TypeKind();

```

File: src/Microsoft.OData.Core/JsonLight/ODataJsonLightParameterDeserializer.cs [hunk 6]
```diff
@@ -216,5 +222,192 @@ internal bool ReadNextParameter(PropertyAndAnnotationCollector propertyAndAnnota
...
+                                switch (parameterTypeKind)
+                                {
+                                    case EdmTypeKind.Primitive:
+                                        IEdmPrimitiveTypeReference primitiveTypeReference = parameterTypeReference.AsPrimitive();
```
Comment (KenitoInc) [inline src/Microsoft.OData.Core/JsonLight/ODataJsonLightParameterDeserializer.cs:291]:
```suggestion
                                        IEdmPrimitiveTypeReference primitiveTypeReference = parameterTypeReference.AsPrimitive();

```

**Review by KenitoInc — APPROVED** [review]
Approved with minor comments

File: src/Microsoft.OData.Core/JsonLight/ODataJsonLightParameterDeserializer.cs (hunk 3 above)
Comment (gathogojr) [reply src/Microsoft.OData.Core/JsonLight/ODataJsonLightParameterDeserializer.cs:29]: Addressed by moving the statement about `ODataJsonLightParameterDeserializer` not supporting reading of property annotations to the body of the delegate

**Pull request merged**